File: linux/argent_utils/selfupdate.py
"""Self-update: fast-forward the checkout, rebuild argent-core, relaunch.

The Linux applet runs straight out of its git checkout (the autostart .desktop
Exec points at ``linux/argent-utils``), so "update" is three steps: fast-forward
the checkout to its upstream, re-run ``scripts/build-core.sh`` to refresh the
installed argent-core binary, and start the launcher again. Relaunching is all
the "reinstall" needed — the singleton is newest-wins (see ``singleton.py``),
so the fresh instance asks the old one to quit the moment it starts.

Everything here is synchronous and shell-based; the Store wraps it in daemon
threads (``refresh_update_status_async`` / ``update_applet_async``) the same
way it wraps the device-allocator installer.
"""

from __future__ import annotations

import os
import subprocess
from pathlib import Path


class UpdateError(RuntimeError):
    """A self-update step failed; str(exc) is the user-facing reason."""


def repo_root() -> Path:
    """The checkout the running applet lives in (env-overridable for tests)."""
    env = os.environ.get("ARGENT_UTILS_SELF_REPO")
    if env:
        return Path(env)
    # This file is <repo>/linux/argent_utils/selfupdate.py, so parents[2] = <repo>.
    return Path(__file__).resolve().parents[2]


def _git(root: Path, *args: str, timeout: float = 120.0) -> str:
    proc = subprocess.run(  # noqa: S603 — fixed argv, no shell
        ["git", "-C", str(root), *args],
        capture_output=True,
        text=True,
        timeout=timeout,
    )
    if proc.returncode != 0:
        lines = (proc.stderr.strip() or proc.stdout.strip()).splitlines()
        raise UpdateError(f"git {args[0]}: {lines[-1] if lines else f'exit {proc.returncode}'}")
    return proc.stdout.strip()


def _upstream(root: Path) -> str:
    """The ref we update to: the branch's upstream, else the default branch."""
    try:
        return _git(root, "rev-parse", "--abbrev-ref", "--symbolic-full-name", "@{upstream}")
    except UpdateError:
        return "origin/main"


def check() -> dict:
    """Fetch origin and report where the checkout stands vs its upstream.

    Never raises (same contract as ``deviceallocator.check``): an unreachable
    remote still yields the local commit plus an ``error`` string.
    """
    root = repo_root()
    out: dict = {
        "root": str(root),
        "commit": None,
        "branch": None,
        "upstream": None,
        "behind": None,
        "ahead": None,
        "error": None,
    }
    try:
        out["commit"] = _git(root, "rev-parse", "--short", "HEAD")
        out["branch"] = _git(root, "rev-parse", "--abbrev-ref", "HEAD")
        _git(root, "fetch", "--quiet", "origin")
        up = _upstream(root)
        out["upstream"] = up
        # left = commits only on HEAD (ahead), right = only on upstream (behind).
        ahead, behind = _git(
            root, "rev-list", "--left-right", "--count", f"HEAD...{up}"
        ).split()
        out["ahead"] = int(ahead)
        out["behind"] = int(behind)
    except (UpdateError, OSError, subprocess.TimeoutExpired, ValueError) as exc:
        out["error"] = str(exc)
    return out


def _has_git_identity(root: Path) -> bool:
    """Whether a committer name+email is configured (a merge commit needs one)."""
    for key in ("user.name", "user.email"):
        proc = subprocess.run(  # noqa: S603 — fixed argv, no shell
            ["git", "-C", str(root), "config", "--get", key],
            capture_output=True, text=True, timeout=30,
        )
        if proc.returncode != 0 or not proc.stdout.strip():
            return False
    return True


def pull() -> str:
    """Integrate the checkout's upstream; returns the resulting short SHA.

    Fast-forwards when the checkout is strictly behind, and creates a merge
    commit when it has diverged (local commits origin doesn't have) — so an
    update still lands when you're *ahead*, which ``--ff-only`` refused to do.

    A *real* conflict is never resolved unattended: the merge is aborted, the
    checkout is left byte-for-byte as it was, and a readable ``UpdateError``
    says it needs a manual merge. Uncommitted local changes still block the
    merge outright — an update must never clobber work in flight.
    """
    root = repo_root()
    if _git(root, "status", "--porcelain", "--untracked-files=no"):
        raise UpdateError("checkout has local changes — commit or stash them first")
    _git(root, "fetch", "--quiet", "origin")
    up = _upstream(root)

    # A diverged checkout needs a merge commit; give the auto-merge a committer
    # identity if the environment has none (a stripped 6AM service env might),
    # but never override the user's own identity when it's configured.
    ident: list[str] = []
    if not _has_git_identity(root):
        ident = ["-c", "user.name=Argent Utils updater",
                 "-c", "user.email=argent-utils@localhost"]
    merge = subprocess.run(  # noqa: S603 — fixed argv, no shell
        ["git", "-C", str(root), *ident, "merge", "--no-edit", up],
        capture_output=True, text=True, timeout=120,
    )
    if merge.returncode != 0:
        blob = "\n".join(p for p in (merge.stdout, merge.stderr) if p).strip()
        # Leave nothing half-merged behind, whatever went wrong. Harmless
        # (just errors) if no merge was actually started.
        subprocess.run(  # noqa: S603 — fixed argv, no shell
            ["git", "-C", str(root), "merge", "--abort"],
            capture_output=True, text=True, timeout=120,
        )
        if "conflict" in blob.lower():
            raise UpdateError(
                "update conflicts with your local commits — merge origin by hand "
                "in the checkout, then update again"
            )
        last = blob.splitlines()[-1] if blob else f"exit {merge.returncode}"
        raise UpdateError(f"git merge: {last}")
    return _git(root, "rev-parse", "--short", "HEAD")


def build_core() -> None:
    """Run the (freshly pulled) checkout's build-core.sh.

    Rebuilds the static argent-core CLI and installs it to
    ``~/.local/share/argent-utils/argent-core`` — the actual "reinstall" step;
    the applet's Python code needs no install, it runs from the checkout.
    """
    root = repo_root()
    script = root / "linux" / "scripts" / "build-core.sh"
    proc = subprocess.run(  # noqa: S603 — our own build script
        ["bash", str(script)],
        cwd=str(root),
        capture_output=True,
        text=True,
        timeout=1800,  # a cold swift release build can take a while
    )
    if proc.returncode != 0:
        lines = (proc.stderr.strip() or proc.stdout.strip()).splitlines()
        raise UpdateError(
            f"build-core.sh: {lines[-1] if lines else f'exit {proc.returncode}'}"
        )


def _state_dir() -> Path:
    return (
        Path(os.environ.get("XDG_STATE_HOME") or Path.home() / ".local" / "state")
        / "argent-utils"
    )


def relaunch(extra_env: dict[str, str] | None = None) -> None:
    """Start the updated launcher detached, logging where autostart logs.

    The new instance's newest-wins singleton terminates this process once it's
    up, so the caller only reports "restarting…" and waits to be replaced.

    ``extra_env`` is merged over the current environment for the child — the 6AM
    updater uses it to hand the relaunched GUI the display env (DISPLAY / Wayland
    / D-Bus) of the tray it's replacing, which a bare systemd service env lacks.
    """
    root = repo_root()
    launcher = root / "linux" / "argent-utils"
    log_dir = _state_dir()
    env = dict(os.environ)
    if extra_env:
        env.update(extra_env)
    try:
        log_dir.mkdir(parents=True, exist_ok=True)
        with (log_dir / "argent-utils.log").open("ab") as log:
            subprocess.Popen(  # noqa: S603 — relaunch ourselves, detached
                ["bash", str(launcher)],
                cwd=str(root),
                start_new_session=True,
                stdin=subprocess.DEVNULL,
                stdout=log,
                stderr=subprocess.STDOUT,
                env=env,
            )
    except OSError as exc:
        raise UpdateError(f"could not relaunch the applet: {exc}") from exc


# MARK: unattended (6AM timer) path

# The display/session vars a relaunched GUI needs but a systemd/launchd service
# env doesn't carry; we lift them off the running tray's process environment.
_DISPLAY_ENV_KEYS = (
    "DISPLAY",
    "WAYLAND_DISPLAY",
    "XAUTHORITY",
    "DBUS_SESSION_BUS_ADDRESS",
    "XDG_RUNTIME_DIR",
    "XDG_SESSION_TYPE",
    "XDG_CURRENT_DESKTOP",
)


def _display_env_of(pid: int) -> dict[str, str]:
    """The display/session env of a running process (via ``/proc/<pid>/environ``)."""
    out: dict[str, str] = {}
    try:
        raw = Path(f"/proc/{pid}/environ").read_bytes()
    except OSError:
        return out
    for entry in raw.split(b"\0"):
        key, sep, val = entry.partition(b"=")
        if sep and key.decode("utf-8", "replace") in _DISPLAY_ENV_KEYS:
            out[key.decode("utf-8", "replace")] = val.decode("utf-8", "replace")
    return out


def _sched_log(message: str) -> None:
    """Append a timestamped line to the auto-update log (best-effort)."""
    from datetime import datetime

    line = f"{datetime.now().isoformat(timespec='seconds')} {message}\n"
    try:
        d = _state_dir()
        d.mkdir(parents=True, exist_ok=True)
        with (d / "autoupdate.log").open("a") as fh:
            fh.write(line)
    except OSError:
        pass


def run_scheduled() -> int:
    """Headless daily update for the 6AM timer. Never raises; returns an exit code.

    Fetches, and if the checkout is behind, merges upstream and rebuilds
    argent-core — then relaunches the tray only if one is actually running (so it
    never pops a GUI on a session that has none). Quiet no-op when already
    current; a conflict or an unreachable origin is logged and left for a human
    rather than retried destructively.
    """
    st = check()
    if st["error"]:
        _sched_log(f"skip: cannot reach origin ({st['error']})")
        return 0
    if not st["behind"]:
        extra = f" ({st['ahead']} local ahead)" if st.get("ahead") else ""
        _sched_log(f"up to date at {st['commit']}{extra}")
        return 0

    _sched_log(f"{st['behind']} behind at {st['commit']} — merging {st['upstream']}")
    try:
        commit = pull()
    except UpdateError as exc:
        # A conflict/dirty tree is not a transient failure to hammer on; wait
        # for the user to sort it, then the next 6AM tick picks it up.
        _sched_log(f"skip: {exc}")
        return 0

    _sched_log(f"merged to {commit} — rebuilding argent-core")
    try:
        build_core()
    except UpdateError as exc:
        _sched_log(f"build failed: {exc}")
        return 1

    from .singleton import SingleInstance

    pid = SingleInstance.running_pid()
    if pid:
        relaunch(_display_env_of(pid))
        _sched_log(f"relaunched running tray (was pid {pid}) onto {commit}")
    else:
        _sched_log(f"updated to {commit} in place (tray not running)")
    return 0

File: Sources/CoMaintainer/SelfUpdate.swift
import Foundation

/// Self-update for the macOS app: fast-forward the checkout, rebuild the `.app`
/// bundle, and relaunch it. The Swift port of the Linux front-end's `selfupdate`
/// module, adapted for the packaged front-end — where Linux relaunches the checkout's
/// launcher, macOS rebuilds `CoMaintainer.app` (via `scripts/build-app.sh`) and `open`s
/// it; the newest-wins singleton (`SingleInstance`) then terminates this instance, so
/// a successful update ends with this process about to be replaced.
///
/// Everything here is synchronous and shell-based; the Store wraps it in detached tasks
/// (`refreshUpdateStatus` / `updateApp`) the same way it wraps the allocator installer.
enum SelfUpdate {
    struct UpdateError: LocalizedError {
        let message: String
        var errorDescription: String? { message }
    }

    /// Where the checkout stands vs its upstream — the payload behind the Settings
    /// "UPDATE" section. Mirrors `selfupdate.check`'s dict.
    struct CheckResult: Equatable {
        var commit: String?
        var branch: String?
        var upstream: String?
        var behind: Int?
        var ahead: Int?
        var error: String?
    }

    private static var root: URL { RepoPaths.root }

    // MARK: - git plumbing

    /// Run `git -C <root> …`, returning trimmed stdout; throws `UpdateError` (last stderr
    /// line) on a non-zero exit. Mirrors `selfupdate._git`.
    @discardableResult
    private static func git(_ args: [String], timeout: TimeInterval = 120) throws -> String {
        let p = Process()
        p.executableURL = URL(fileURLWithPath: "/usr/bin/git")
        p.arguments = ["-C", root.path] + args
        let out = Pipe(), err = Pipe()
        p.standardOutput = out
        p.standardError = err
        do { try p.run() } catch {
            throw UpdateError(message: "git \(args.first ?? ""): \(error.localizedDescription)")
        }
        let watchdog = DispatchWorkItem { if p.isRunning { p.terminate() } }
        DispatchQueue.global().asyncAfter(deadline: .now() + timeout, execute: watchdog)
        let outData = out.fileHandleForReading.readDataToEndOfFile()
        let errData = err.fileHandleForReading.readDataToEndOfFile()
        p.waitUntilExit()
        watchdog.cancel()
        let stdout = String(data: outData, encoding: .utf8) ?? ""
        if p.terminationStatus != 0 {
            let stderr = String(data: errData, encoding: .utf8) ?? ""
            let lines = (stderr.isEmpty ? stdout : stderr)
                .split(whereSeparator: \.isNewline)
            let detail = lines.last.map(String.init) ?? "exit \(p.terminationStatus)"
            throw UpdateError(message: "git \(args.first ?? ""): \(detail)")
        }
        return stdout.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    /// The ref we update to: the branch's upstream, else `origin/main`.
    private static func upstream() -> String {
        (try? git(["rev-parse", "--abbrev-ref", "--symbolic-full-name", "@{upstream}"])) ?? "origin/main"
    }

    // MARK: - public surface

    /// Fetch origin and report where the checkout stands vs its upstream. Never throws —
    /// an unreachable remote still yields the local commit plus an `error` string.
    static func check() -> CheckResult {
        var out = CheckResult()
        guard RepoPaths.checkoutPresent else {
            out.error = "no checkout at \(root.path) — set CO_MAINTAINER_SELF_REPO"
            return out
        }
        do {
            out.commit = try git(["rev-parse", "--short", "HEAD"])
            out.branch = try git(["rev-parse", "--abbrev-ref", "HEAD"])
            try git(["fetch", "--quiet", "origin"])
            let up = upstream()
            out.upstream = up
            // left = commits only on HEAD (ahead), right = only on upstream (behind).
            let counts = try git(["rev-list", "--left-right", "--count", "HEAD...\(up)"])
                .split(whereSeparator: \.isWhitespace)
            if counts.count == 2 {
                out.ahead = Int(counts[0])
                out.behind = Int(counts[1])
            }
        } catch let e as UpdateError {
            out.error = e.message
        } catch {
            out.error = "\(error)"
        }
        return out
    }

    /// Whether a committer name+email is configured (a merge commit needs one).
    private static func hasGitIdentity() -> Bool {
        for key in ["user.name", "user.email"] {
            let v = (try? git(["config", "--get", key])) ?? ""
            if v.isEmpty { return false }
        }
        return true
    }

    /// Integrate the checkout's upstream; returns the resulting short SHA. Fast-forwards
    /// when strictly behind, and creates a merge commit when the checkout has diverged
    /// (local commits origin doesn't have) — so an update still lands when you're *ahead*,
    /// which `--ff-only` refused to do. A real conflict is never resolved unattended: the
    /// merge is aborted, the checkout left as it was, and a readable error says it needs a
    /// manual merge. Uncommitted local changes still block outright. Mirrors `selfupdate.pull`.
    static func pull() throws -> String {
        let dirty = try git(["status", "--porcelain", "--untracked-files=no"])
        if !dirty.isEmpty {
            throw UpdateError(message: "checkout has local changes — commit or stash them first")
        }
        try git(["fetch", "--quiet", "origin"])
        let up = upstream()
        // Give the auto-merge a committer identity if the environment has none (a stripped
        // launchd service env might), but never override the user's own when it's set.
        let ident = hasGitIdentity()
            ? []
            : ["-c", "user.name=Co-Maintainer updater", "-c", "user.email=co-maintainer@localhost"]
        do {
            try git(ident + ["merge", "--no-edit", up])
        } catch let e as UpdateError {
            // Leave nothing half-merged behind, whatever went wrong.
            _ = try? git(["merge", "--abort"])
            if e.message.lowercased().contains("conflict") {
                throw UpdateError(message: "update conflicts with your local commits — merge origin "
                    + "by hand in the checkout, then update again")
            }
            throw e
        }
        return try git(["rev-parse", "--short", "HEAD"])
    }

    /// Rebuild `CoMaintainer.app` from the (freshly pulled) checkout via
    /// `scripts/build-app.sh`. Run through a login shell so the Swift toolchain is on
    /// PATH even when the app was started from launchd with a minimal environment.
    static func rebuild() throws {
        let script = root.appendingPathComponent("scripts/build-app.sh").path
        guard FileManager.default.fileExists(atPath: script) else {
            throw UpdateError(message: "scripts/build-app.sh not found in \(root.path)")
        }
        let p = Process()
        p.executableURL = URL(fileURLWithPath: "/bin/bash")
        p.arguments = ["-lc", "exec \(shellQuote(script))"]
        p.currentDirectoryURL = root
        let err = Pipe()
        p.standardOutput = FileHandle.nullDevice
        p.standardError = err
        do { try p.run() } catch {
            throw UpdateError(message: "build-app.sh: \(error.localizedDescription)")
        }
        // A cold release build can take a while.
        let watchdog = DispatchWorkItem { if p.isRunning { p.terminate() } }
        DispatchQueue.global().asyncAfter(deadline: .now() + 1800, execute: watchdog)
        let errData = err.fileHandleForReading.readDataToEndOfFile()
        p.waitUntilExit()
        watchdog.cancel()
        if p.terminationStatus != 0 {
            let lines = (String(data: errData, encoding: .utf8) ?? "").split(whereSeparator: \.isNewline)
            throw UpdateError(message: "build-app.sh: \(lines.last.map(String.init) ?? "exit \(p.terminationStatus)")")
        }
    }

    /// Launch the freshly-built bundle detached; its newest-wins singleton terminates
    /// this instance once it's up, so the caller only reports "restarting…" and waits to
    /// be replaced. Mirrors `selfupdate.relaunch`.
    static func relaunch() throws {
        let app = root.appendingPathComponent("CoMaintainer.app")
        guard FileManager.default.fileExists(atPath: app.path) else {
            throw UpdateError(message: "CoMaintainer.app not found after rebuild")
        }
        let p = Process()
        p.executableURL = URL(fileURLWithPath: "/usr/bin/open")
        p.arguments = ["-n", app.path]
        do { try p.run() } catch {
            throw UpdateError(message: "could not relaunch the app: \(error.localizedDescription)")
        }
        p.waitUntilExit()
        if p.terminationStatus != 0 {
            throw UpdateError(message: "open CoMaintainer.app exited \(p.terminationStatus)")
        }
    }

    private static func shellQuote(_ s: String) -> String {
        "'" + s.replacingOccurrences(of: "'", with: "'\\''") + "'"
    }

    // MARK: - unattended (6AM launchd) path

    /// Headless daily update for the launchd 6AM job. Never throws; returns an exit code.
    ///
    /// Fetches, and if behind, merges upstream and rebuilds — then relaunches the app only
    /// if one is actually running (so it never pops a menu-bar app onto a login session that
    /// isn't showing one). Quiet no-op when already current; a conflict or unreachable origin
    /// is logged and left for a human rather than retried destructively. Mirrors
    /// `selfupdate.run_scheduled`.
    static func runScheduled() -> Int32 {
        let st = check()
        if let e = st.error { schedLog("skip: cannot reach origin (\(e))"); return 0 }
        guard let behind = st.behind, behind > 0 else {
            let extra = (st.ahead ?? 0) > 0 ? " (\(st.ahead!) local ahead)" : ""
            schedLog("up to date at \(st.commit ?? "?")\(extra)")
            return 0
        }
        schedLog("\(behind) behind at \(st.commit ?? "?") — merging \(st.upstream ?? "origin/main")")
        let commit: String
        do {
            commit = try pull()
        } catch {
            schedLog("skip: \((error as? LocalizedError)?.errorDescription ?? "\(error)")")
            return 0
        }
        schedLog("merged to \(commit) — rebuilding the app")
        do {
            try rebuild()
        } catch {
            schedLog("build failed: \((error as? LocalizedError)?.errorDescription ?? "\(error)")")
            return 1
        }
        if SingleInstance.isRunning() {
            do {
                try relaunch()
                schedLog("relaunched running app onto \(commit)")
            } catch {
                schedLog("relaunch failed: \((error as? LocalizedError)?.errorDescription ?? "\(error)")")
                return 1
            }
        } else {
            schedLog("updated to \(commit) in place (app not running)")
        }
        return 0
    }

    /// Append a timestamped line to the auto-update log (best-effort).
    private static func schedLog(_ message: String) {
        let dir = FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent("Library/Logs")
        let url = dir.appendingPathComponent("co-maintainer-autoupdate.log")
        let line = "\(ISO8601DateFormatter().string(from: Date())) \(message)\n"
        guard let data = line.data(using: .utf8) else { return }
        if let fh = try? FileHandle(forWritingTo: url) {
            defer { try? fh.close() }
            _ = try? fh.seekToEnd()
            try? fh.write(contentsOf: data)
        } else {
            try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
            try? data.write(to: url)
        }
    }
}
